[3/N] Remove events from Flight Recorder

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141737
* #141712
* #141648

Initially we just want to generalize `CUDAEvent` to `c10::Event` to make FR device agnostic. But we realized that the conversion is not available today without change in `aten::cuda::CUDAEvent`. Thus we consider an alternative which removes Events from FR.

### How are events used in FR today? Are they replaceable?
In two places:
1. To calculate work duration in `retire_id`;
2. To `update_state` of an entry during `dump`, where the states relate to start time & complete time.

Re 1:
Watchdog is calling `retire_id` anyway. So it doesn't hurt for watchdog to calculate the duration and give it to FR.
This PR changes `retire_id` signature from:
`retire_id(id, bool compute_duration)` to
`retire_id(id, std::optional duration)`.

Re 2:
The `update_state` call seems to be for a case where watchdog hanged and did not update the state of an entry in time. FR updates it later when dumping the trace. Personally I doubt this works because the hang of watchdog means CUDA is already obstructed and there is little chance FR can do it instead. Moreover, it introduced CUDA calls in the dump behavior (executed by the heartbeat thread), making it hang'able too which is not desired.

### Other Considerations
3. **If NCCL comm is aborting, no "complete" information is reliable afterwards**. Thus we prefer watchdog to explicitly mark complete than FR doing so in the background or near the end of the program (dumping).

4. It is a higher bar to ask **all** accelerators to never hang in event query, once FR is generalized. Or at least, there are unknowns. Thus this PR takes a conservative mental model.

5. Today FR keeps a raw pointer of the events, which poses a leak to the event cache that is based on shared_ptr?

Per the above considerations, this PR removes events from FR.

To compensate for start/complete/duration info, this PR adds `markStart` method and provides duration to `retire_id`.

CC: @c-p-i-o @fduwjj @wconstab 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[3/N] Remove events from Flight Recorder

[ghstack-poisoned]

kwen2501 committed:
Update on "[3/N] Remove events from Flight Recorder"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[3/N] Remove events from Flight Recorder"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2115,6 +2118,7 @@ void ProcessGroupNCCL::watchdogHandler() {
       if (work.isCompleted()) {
         // Work status logging for desync debug
         desyncDebugger_.logWorkEnd(work);
+        FlightRecorder::get()->markEnd(work.trace_id_, work.getDuration());
```
This is bad. Because when watchdog is hang, FR will dump inaccurate information because we didn't mark collectives to be "complete".

kwen2501 replied:
Yeah, I understand this. I gave a trade-off analysis in "Re 2".

kwen2501 replied:
I also added this: 
if NCCL comm is aborting, no "complete" information is reliable. Thus we prefer watchdog to explicitly mark entry as complete than FR doing so in the background or near the end of the program.

kwen2501 replied:
Exactly, I consider it a higher bar to ask all accelerators to never hang in event query.

kwen2501 committed:
Update on "[3/N] Remove events from Flight Recorder"


Initially we just want to generalize `CUDAEvent` to `c10::Event` to make FR device agnostic. But we realized that the conversion is not available today without change in `aten::cuda::CUDAEvent`. Thus we consider an alternative which removes Events from FR.

### How are events used in FR today? Are they replaceable?
In two places:
1. To calculate work duration in `retire_id`;
2. To `update_state` of an entry during `dump`, where the states relate to start time & complete time.

Re 1:
Watchdog is calling `retire_id` anyway. So it doesn't hurt for watchdog to calculate the duration and give it to FR.

Re 2:
The `update_state` call seems to be for a case where watchdog hanged and did not update the state of an entry in time. Thus FR updates it when dumping the trace. Personally I doubt this works because the hang of watchdog means CUDA is already obstructed and there is little chance FR can update the state. Moreover, it introduced CUDA calls in the dump behavior (executed by the heartbeat thread), making it hang'able too which is not desired.

### Other Considerations
- **If NCCL comm is aborting, no "complete" information is reliable afterwards**. Thus we prefer watchdog to explicitly mark complete than FR doing so in the background or near the end of the program.

- It is a higher bar to ask **all** accelerators to never hang in event query, once FR is generalized. Or at least, there are unknowns here. Thus this PR takes a conservative mental model.

Per the above considerations, this PR removes events from FR, **to make it more "hang-free"**.

To compensate for start/complete/duration info, this PR adds `markStart` method and provides duration to `retire_id`.

CC: c-p-i-o fduwjj wconstab 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

wconstab commented:
I'm excited to take a closer look at this after the holiday. I don't want to rush this one. Thanks for doing this though!

kwen2501 committed:
Update on "[3/N] Remove events from Flight Recorder"


Initially we just want to generalize `CUDAEvent` to `c10::Event` to make FR device agnostic. But we realized that the conversion is not available today without change in `aten::cuda::CUDAEvent`. Thus we consider an alternative which removes Events from FR.

### How are events used in FR today? Are they replaceable?
In two places:
1. To calculate work duration in `retire_id`;
2. To `update_state` of an entry during `dump`, where the states relate to start time & complete time.

Re 1:
Watchdog is calling `retire_id` anyway. So it doesn't hurt for watchdog to calculate the duration and give it to FR.
This PR changes `retire_id` signature from:
`retire_id(id, bool compute_duration)` to
`retire_id(id, std::optional duration)`.

Re 2:
The `update_state` call seems to be for a case where watchdog hanged and did not update the state of an entry in time. FR updates it later when dumping the trace. Personally I doubt this works because the hang of watchdog means CUDA is already obstructed and there is little chance FR can update the state. Moreover, it introduced CUDA calls in the dump behavior (executed by the heartbeat thread), making it hang'able too which is not desired.

### Other Considerations
3. **If NCCL comm is aborting, no "complete" information is reliable afterwards**. Thus we prefer watchdog to explicitly mark complete than FR doing so in the background or near the end of the program (dumping).

4. It is a higher bar to ask **all** accelerators to never hang in event query, once FR is generalized. Or at least, there are unknowns. Thus this PR takes a conservative mental model.

5. Today FR keeps a raw pointer of the events, which poses a leak to the event cache that is based on shared_ptr?

Per the above considerations, this PR removes events from FR.

To compensate for start/complete/duration info, this PR adds `markStart` method and provides duration to `retire_id`.

CC: c-p-i-o fduwjj wconstab 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o 

[ghstack-poisoned]

c-p-i-o commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -316,51 +300,14 @@ std::optional<FlightRecorder::Entry> FlightRecorder::getEntry(
-    bool compute_duration) {
-  if (!enabled_ || !id) {
+    std::optional<float> duration) {
+  auto entry = getEntry(id);
+  if (!entry)
     return;
```
Is there a potential race here - assuming multiple threads are working on the same entry?
For example, we do `getEntry` and it returns slot `0` entry.
But, before we get line `308`, that slot gets overwritten by the `record` function?

kwen2501 replied:
Good point. I should move the `lock_guard` in line 307 to the top. Thanks.

c-p-i-o replied:
I think will deadlock if you just move it up - since `getEntry` also attempts to grab the same mutex_. Maybe need some refactoring to pass the lock? Or remove lock from getEntry and assume that it can only be called with a lock held?

kwen2501 replied:
Yeah, the lock in getEntry should be removed. Otherwise we will need to use a recursive mutex.

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -289,25 +278,20 @@ std::vector<FlightRecorder::Entry> FlightRecorder::dump_entries() {
     return std::nullopt;
   }
 
   std::unique_lock<std::mutex> guard(mutex_);
```
Yeah, we should unify the code. In most cases they are the same. `unique_lock` will release the mutex when it goes out of scope.

c-p-i-o replied:
I think will deadlock if you just move it up - since `getEntry` also attempts to grab the same `mutex_`. Maybe need some refactoring to pass the lock? Or remove lock from `getEntry` and assume that it can only be called with a lock held?

wconstab commented:
> Today FR keeps a raw pointer of the events. Does it pose a leak to the event cache which is based on shared_ptr?

IIRC, no. The reason is that the existing FR is designed to only keep raw ptrs to events that are still owned by live Work objs.  Once the work obj goes out of scope and returns the event to the cache, the FR event should have also been retired.  IF that is not true, then we'd have a leak.

wconstab commented:
Was trying to figure out why we had things the way they were, and I asked @zdevito, and we agreed that cudaEventQuery is special, it will not hang when the rest of the cuda APIs hang.  (We even have a [unit test that verifies this](https://github.com/pytorch/pytorch/blob/main/test/distributed/test_c10d_nccl.py#L4913)).

So in the existing code, the design principle was 
(1) - calculating timing is risky, it could hang.  The hang is exposed to the watchdog, but if the watchdog hangs the monitor will catch it and dump the recorder.
(2) - 'dump()' may never hang (by design) - and cudaEventQuery itself is all that dump() calls, which itself does not hang, so this is fine.

For the new code, the parts relating to (1) seem fine to me.  But the parts relating to (2) may need more thought.  We don't want to make 'dump()' exposed to a hang risk, and we want to ensure that 'dump()' is giving updated/latest state info even if the watchdog loop did hang.

kwen2501 commented:
Thanks for the detailed feedback.

> we want to ensure that 'dump()' is giving updated/latest state info even if the watchdog loop did hang

I agree this is the drawback of the PR as is. And I made it a draft till we have a solution.

I considered a couple pros and cons:

1. When users call dump, we wake up the watchdog thread to loop another round through the work queue. 
This will work when watchdog is healthy. But will not work if watchdog already hangs.

2. We can let FR have pointer to events, and query them when FR dumps. The only problem is that FR does not have knowledge of the comm state, hence cannot distinguish whether a kernel really completes or is killed. (Ideally, we should have states like "scheduled", "completed", "killed". Only `ProcessGroupNCCL` has such knowledge.)

wconstab commented:
> 2. We can let FR have pointer to events, and query them when FR dumps. The only problem is that FR does not have knowledge of the comm state, hence cannot distinguish whether a kernel really completes or is killed. (Ideally, we should have states like "scheduled", "completed", "killed". Only ProcessGroupNCCL has such knowledge.)

This is a good point- it would be nice if there were an 'aborted' state.  We could add that as a follow up.  We could expose a kwarg for 'isAborted' to the 'dump()' api (and maybe also the retire api?) so the FR internals don't need to check the PGNccl status, but they can still know if there has been an abort.  (Can you abort just one PG and not all the PGs though?  It's hard to convey this information cleanly)

But what do we do about 'dump' not hanging for this PR?  Maybe what you want to do is move the eventQuery call outside of dump() and instead put another loop inside heartbeat monitor that loops over the remaining worklist and retires all the FR ids.  then dump() should assume/assert that all works are retired before it is called.
